src/Microsoft.Extensions.Options/OptionsCache.cs

```diff
@@ -1,49 +1,65 @@
-// Copyright (c) .NET Foundation. All rights reserved.
+﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
-using System.Collections.Generic;
-using System.Threading;
+using System.Collections.Concurrent;
 
 namespace Microsoft.Extensions.Options
 {
-    internal class OptionsCache<TOptions> where TOptions : class, new()
+    /// <summary>
+    /// Used to cache TOptions instances.
+    /// </summary>
+    /// <typeparam name="TOptions">The type of options being requested.</typeparam>
+    public class OptionsCache<TOptions> : IOptionsCache<TOptions> where TOptions : class
     {
-        private readonly Func<TOptions> _createCache;
-        private object _cacheLock = new object();
-        private bool _cacheInitialized;
-        private TOptions _options;
-        private IEnumerable<IConfigureOptions<TOptions>> _setups;
+        private readonly ConcurrentDictionary<string, Lazy<TOptions>> _cache = new ConcurrentDictionary<string, Lazy<TOptions>>(StringComparer.Ordinal);
 
-        public OptionsCache(IEnumerable<IConfigureOptions<TOptions>> setups)
+        public virtual TOptions GetOrAdd(string name, Func<TOptions> createOptions)
         {
-            _setups = setups;
-            _createCache = CreateOptions;
+            if (name == null)
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+            if (createOptions == null)
+            {
+                throw new ArgumentNullException(nameof(createOptions));
+            }
+            return _cache.GetOrAdd(name, new Lazy<TOptions>(createOptions)).Value;
```

Comment:
That's not really about being thread-safe; it's about the two different patterns for initializing something in a thread-safe way. The one that is generally preferred is that if the object you are initializing is not yet initialized, then you don't try to prevent two things from creating it, but ultimately only one object wins and gets stored. This is usually fine. There are occasions where the initialization code itself cannot run concurrently (usually indicates a different design issue with the initialization code) or it is important that it only ever run one because it has side effects. It does not seem to be that either of these things should be necessary here.